### ISSUE TITLE
Flex Attention breaks in certain cases when used with a learned bias

### ISSUE THREAD

**Issue by vladkvit** [pytorch/pytorch#157677]
### 🐛 Describe the bug

I have a neural net that uses F.scaled_dot_product_attention together with a learned bias. I am trying to switch to Flex Attention. I am running into strange issues where Flex Attention gives me an assertion. Things like commenting out a print or removing an output layer seem to make it work. 

Below is the smallest reproducible code I could put together. Note the "Switch this line" comment, "Uncomment this line" and "Comment this line" comments.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

from torch.nn.attention.flex_attention import (
    flex_attention,
)

batch_size = 2
sequence_length = 64
d_model = 256
n_head = 4


class BiasGen(nn.Module):
    def __init__(self, d_model: int, nhead: int, context_length: int, dtype=None):
        super().__init__()
        self.context_length = context_length
        self.n_head = nhead

        # Simple projection to generate attention bias
        self.bias_proj = nn.Linear(
            d_model, nhead * context_length * context_length, dtype=dtype
        )

    def forward(self, src: torch.Tensor):
        BATCH_SIZE = src.shape[0]

        # Use the first token to generate bias for the entire sequence
        first_token = src[:, 0, :]  # [B, d_model]

        # Project to attention bias
        bias_flat = self.bias_proj(
            first_token
        )  # [B, nhead * context_length * context_length]

        # Reshape to attention bias format
        bias = bias_flat.view(
            BATCH_SIZE, self.n_head, self.context_length, self.context_length
        )

        return bias


class ChessAttention(nn.Module):
    def __init__(
        self,
        d_model,
        num_heads,
    ):
        super(ChessAttention, self).__init__()
        assert d_model % num_heads == 0, "embed_dim is indivisible by num_heads"
        self.num_heads = num_heads
        self.d_model = d_model
        self.qk_dim = d_model
        self.qk_head_dim = self.qk_dim // num_heads
        self.v_head_dim = self.d_model // num_heads

        self.q_proj = nn.Linear(d_model, self.qk_dim)
        self.k_proj = nn.Linear(d_model, self.qk_dim)
        self.v_proj = nn.Linear(d_model, self.d_model)

        self.out_proj = nn.Linear(self.d_model, d_model)

        self.bias_gen = BiasGen(d_model, n_head, sequence_length)

    def forward(
        self,
        x: torch.Tensor,
    ):
        batch_size, seq_len, embed_dim = x.size()

        attn_bias = self.bias_gen(x)

        # Project Q, K, V. (b, num_tokens, embed_dim) --> (b, num_tokens, qkv_dim)
        q = self.q_proj(x)
        k = self.k_proj(x)
        v = self.v_proj(x)

        # (batch_size, num_tokens, qkv_dim) -> (batch_size, num_tokens, num_heads, head_dim)
        q = q.view(batch_size, seq_len, self.num_heads, self.qk_head_dim)
        k = k.view(batch_size, seq_len, self.num_heads, self.qk_head_dim)
        v = v.view(batch_size, seq_len, self.num_heads, self.v_head_dim)

        # SDPA requires batch, head, seq_len, head_emb_dim
        q = q.permute(0, 2, 1, 3)
        k = k.permute(0, 2, 1, 3)
        v = v.permute(0, 2, 1, 3)

        print("DEBUG----", attn_bias.shape, q.shape, k.shape, v.shape, x.shape)

        attn_bias_contiguous = attn_bias.contiguous()

        def bias_mod(score, b, h, q_idx, kv_idx):
            return score + attn_bias_contiguous[b, h, q_idx, kv_idx]

        ### SWITCH THESE LINES AND THINGS WORK ###
        attn_output = flex_attention(q, k, v, score_mod=bias_mod)
        # attn_output = F.scaled_dot_product_attention(q, k, v, attn_mask=attn_bias)

        # fix intellisense nit
        assert isinstance(attn_output, torch.Tensor)

        ### UNCOMMENT THIS LINE AND THINGS WORK ###
        # print(attn_output.shape)

        # Concatenate heads and project to output dimension
        attn_output = (
            attn_output.transpose(1, 2)
            .contiguous()
            .view(batch_size, seq_len, self.d_model)
        )

        ### COMMENT THIS LINE OUT AND THINGS WORK ###
        attn_output = self.out_proj(attn_output)  # (batch_size, seq_len, d_model)

        return attn_output


@torch.compile
def run(attention_model, device):
    random_input = torch.randn(batch_size, sequence_length, d_model, device=device)
    random_target = torch.randn(batch_size, sequence_length, d_model, device=device)

    output = attention_model(random_input)

    loss = torch.nn.functional.mse_loss(output, random_target)

    print(f"Input shape: {random_input.shape}")
    print(f"Output shape: {output.shape}")

    loss.backward()

    print("Backpropagation completed successfully!")

    return loss


def init():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    attention_model = ChessAttention(d_model=d_model, num_heads=n_head)
    attention_model.to(device)
    attention_model.train()
    loss = run(attention_model, device)
    print(f"Loss: {loss.item():.6f}")
    return loss


if __name__ == "__main__":
    loss = init()
```

And here is the stack trace:
```
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\autograd\function.py", line 307, in apply
    return user_fn(self, *args)
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_functorch\_aot_autograd\runtime_wrappers.py", line 2111, in backward
    return impl_fn()
           ^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_functorch\_aot_autograd\runtime_wrappers.py", line 2097, in impl_fn
    out = CompiledFunction._backward_impl(ctx, all_args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_functorch\_aot_autograd\runtime_wrappers.py", line 2186, in _backward_impl
    CompiledFunction.compiled_bw = aot_config.bw_compiler(
                                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_functorch\aot_autograd.py", line 479, in __call__
    return self.compiler_fn(gm, example_inputs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_dynamo\backends\common.py", line 72, in _wrapped_bw_compiler
    return disable(disable(bw_compiler_fn)(*args, **kwargs))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_dynamo\eval_frame.py", line 838, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_utils_internal.py", line 97, in wrapper_function
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_inductor\compile_fx.py", line 2013, in bw_compiler
    return inner_compile(
           ^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_inductor\compile_fx.py", line 628, in compile_fx_inner
    return wrap_compiler_debug(_compile_fx_inner, compiler_name="inductor")(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_dynamo\repro\after_aot.py", line 124, in debug_wrapper
    inner_compiled_fn = compiler_fn(gm, example_inputs)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_inductor\compile_fx.py", line 735, in _compile_fx_inner
    mb_compiled_graph = fx_codegen_and_compile(
                        ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_inductor\compile_fx.py", line 1295, in fx_codegen_and_compile
    return scheme.codegen_and_compile(gm, example_inputs, inputs_to_check, graph_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_inductor\compile_fx.py", line 1119, in codegen_and_compile
    graph.run(*example_inputs)
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_inductor\graph.py", line 877, in run
    return super().run(*args)
           ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\fx\interpreter.py", line 171, in run
    self.env[node] = self.run_node(node)
                     ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_inductor\graph.py", line 1612, in run_node
    result = ir.ExternKernel.require_exact_strides(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_inductor\ir.py", line 5411, in require_exact_strides
    return cls.require_strides(
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_inductor\ir.py", line 5308, in require_strides
    and significant_strides_equal(
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Vlad\AppData\Roaming\Python\Python311\site-packages\torch\_inductor\ir.py", line 417, in significant_strides_equal
    assert len(shape) == len(strides1) and len(strides1) == len(strides2)
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

### Versions

Collecting environment information...
PyTorch version: 2.7.1+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro (10.0.26100 64-bit)
GCC version: Could not collect
Clang version: 19.1.4
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.26100-SP0
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 576.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Name: AMD Ryzen 7 7800X3D 8-Core Processor
Manufacturer: AuthenticAMD
Family: 107
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 4201
MaxClockSpeed: 4201
L2CacheSize: 8192
L2CacheSpeed: None
Revision: 24834

Versions of relevant libraries:
[pip3] flake8==7.1.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] onnx==1.16.0
[pip3] onnxruntime==1.20.1
[pip3] pytorch-ignite==0.5.1
[pip3] pytorch-lightning==2.1.4
[pip3] pytorch-metric-learning==2.3.0
[pip3] rapidocr-onnxruntime==1.3.24
[pip3] torch==2.7.1+cu128
[pip3] torch_geometric==2.4.0
[pip3] torch-tb-profiler==0.4.3
[pip3] torchao==0.7.0+cpu
[pip3] torchaudio==2.7.1+cu128
[pip3] torchmetrics==1.2.1
[pip3] torchtune==0.5.0
[pip3] torchvision==0.22.1+cu128
[pip3] triton-windows==3.3.0.post19
[conda] Could not collect

**Comment by Aceticia** [comment]
I'm getting a similar error, although my code is significantly more complex and it's currently difficult to provide a small repro.

**Comment by jansel** [comment]
My agent was unable to reproduce this on 34424f27313fbcddaafe4a1a855000f17e05a260. Reopen if you see this on the latest PyTorch nightly
